### ISSUE TITLE
Module functions can access uninitialized main.swift globals 

### ISSUE THREAD

**Issue by tristanlabelle** [apple/swift#69535]
**Description**
Global variables used in `main.swift` are visible in functions of the module, which allows their access while uninitialized

**Steps to reproduce**
```swift
class Foo { let message: String = "oh no" }
func getFoo() -> Foo { foo }
let foo = getFoo()
print(foo.message)
```

<img width="617" alt="image" src="https://github.com/apple/swift/assets/2314287/24b2fb9d-a84b-4d3e-9932-9b3a91193b53">

This happens as well if `getFoo()` is moved to a different file in the same module.

**Expected behavior**
The code should fail to compile because `getFoo()` accesses an uninitialized `foo`. More generally, variables in `main.swift` should probably be treated like local variables in a function rather than like global variables.

**Environment**
- Swift compiler version info:
```
compnerd.org Swift version 5.11-dev (LLVM ee13eccb18b0dc4, Swift a94d22e017bd6a4)
Target: x86_64-unknown-windows-msvc
(built from main on 2023-10-27)
```
- Xcode version info: N/A
- Deployment target: Windows 11

**Comment by atrick** [comment]
Any global initializer that refers to the same global attempts to reenter to the initializer. Since the initializer can only be called once, subsequent attempts crash inside libdispatch.dylib_dispatch_once_wait.

I think the compiler should at least attempt to recognize cyclic initialization and report a compile-time error.

**Comment by eeckstein** [comment]
Yeah, we could try to diagnose this error. Though it would be non-trivial because it needs interprocedural analysis.

Also note that global variables in non parse-as-library mode are broken in general because you can access them before initialization - even without such a cycle. In non parse-as-library mode this test program crashes with a segfault. In parse-as-library mode it crashes "as expected" with a trap.
